### ISSUE TITLE
partial validation support

### ISSUE THREAD

**Issue by mbroadst** [ajv-validator/ajv#211]
Hi there, another question here that perhaps stretches the intention of the module! I was wondering if you would consider adding support for so called "partial" validation. Suppose I have the following schema:

```
{
  "type": "object",
  "properties": {
    "first": { "type": "string" },
    "second": {
     "type": "object",
     "properties": {
       "some": { "type": "string" },
       "otherProperties": { "type": "string" }
     }
   }
  },
  "required": [ "first", "second" ]
}
```

And I use that as a schema for an entire record that I provide CRUD actions for. I might have a case where I want to support a `PATCH` (or even just update via `PUT`) to update part of the document, and still want to validate the data however without submitting _all_ the data. Currently submitting a payload of `{ second: { some: 'hello', otherProperties: 'world!' } }` would fail validation because `first` is not present.

**Comment by epoberezkin** [comment]
That's definitely out of scope. This logic should be in the application and to avoid duplication of code in schemas I would use templates with conditionals (or some other mechanism) to generate all the schemas I need from a single template (either at build or at bootstrap time).

**Comment by mbroadst** [comment]
@epoberezkin sorry it took a while to get back to this here. So my current use case for partial validation involves using ajv for form validation in a web app (using aurelia.io). Without the ability to validate a single property in a schema it is difficult to provide inline errors as the user is filling out the form. It seems like there are two options here as a result:
1. run the validator each time a field is modified, make sure to validate with `allErrors: true`, and then match the resulting error object with the corresponding field element
2. only validate the form once, before submit, and report the errors all at the same time

The first option seems like it would get the job done, but it also seems like a lot of validation overhead. There is another possible option: create schemas for each field, but that seems a bit much.

Is the first option your proposed solution for this use case?

**Comment by epoberezkin** [comment]
I was suggesting creating separate schemas for each field - the main schema can be used to extract them programmatically or you can define them separately (and potentially construct main schema by merging them).

The first approach you mention (validate the whole form) can be also justified if some of your fields dependent on the other (e.g. mutually exclusive form subsections toggled by checkbox or password/confirm password fields).

You can also use $ref to extract these partial schemas, e.g. `{ $ref: 'schema.json#/properties/first' }` or `{ $ref: 'schema.json#/properties/second/properties/some' }` - it is part of the standard.

I don't plan to add anything else to support it - I think the approaches above are more than sufficient.

**Comment by mbroadst** [comment]
@epoberezkin lets say I have the following schema:

```
{
    type: 'object',
    properties: {
      firstName: { type: 'string' },
      lastName: { type: 'string' },
      email: { type: 'string', format: 'email' }
    },
    required: [ 'firstName', 'lastName', 'email' ]
  }
```

and for the sake of partial validation, I am splitting out each of the `properties` definitions into individual schemas - how would I maintain the concept of the value being `required`? This is a valid schema on its own `{ type: 'string' }`, but there doesn't seem to be a way to mark it required in isolation (outside of the object definition)

**Comment by mbroadst** [comment]
so I tried your suggestion here: https://github.com/epoberezkin/ajv/issues/272#issuecomment-239605138, in what I _think_ is the proper form:

```
let schema = {
  id: 'someSchema',
  type: 'object',
  properties: {
    firstName: { type: 'string' },
    lastName: { type: 'string' },
    email: { type: 'string', format: 'email' }
  },
  required: [ 'firstName', 'lastName', 'email' ]
};

const Ajv = require('ajv');
let ajv = Ajv();
let validate = ajv.compile(schema);
let valid = validate({ $ref: 'someSchema#firstName' }, 42);
if (!valid) console.log(validate.errors);
```

But the result is:

```
mbroadst@retinoid:test-ajv$ node test3.js
[ { keyword: 'required',
    dataPath: '42',
    schemaPath: '#/required',
    params: { missingProperty: 'firstName' },
    message: 'should have required property \'firstName\'' } ]
```

**Comment by epoberezkin** [comment]
`someSchema#firstName` doesn't point to anything in this case. You either need to use `someSchema#/properties/firstName` or to add `"id": "#firstName"` inside this subschema.

Your code is also wrong, you can't pass schema to a compiled validating function, you can though do `ajv.validate({ $ref: 'someSchema#/properties/firstName' }, 42)`

To have required constrain you need to split your whole schema into multiple subschemas and validate the whole data object rather than its properties:

``` json
{
  "id": "someSchema",
  "type": "object",
  "allOf": [
    {
      "id": "#firstName",
      "properties": {
        "firstName": { "type": "string" }
      },
      "required": [ "firstName" ]
    },
    {
      "id": "#lastName",
      "properties": {
        "lastName": { "type": "string" }
      },
      "required": [ "lastName" ]
    },
    {
      "id": "#email",
      "properties": {
        "email": { "type": "string", "format": "email" }
      },
      "required": [ "email" ]
    }
  ]
}
```

Now you can validate `ajv.validate({ $ref: 'someSchema#firstName' }, { firstName: 42 })`. And you can validate the whole object as well.

Alternatively, as I wrote previously, you can use templates to generate multiple schemas. For example using one doT template:

```
{
  "id": "someSchema",
  "type": "object",
  "properties": {
    {{~ it.properties:p:i }}
      {{?i}},{{?}}
      "{{= p.name }}": {{= JSON.stringify(p.schema) }}
    {{~}}
  },
  "required": [
    {{~ it.properties:p:i }}
      {{?i}},{{?}}
      "{{= p.name }}"
    {{~}}
  ]
}
```

you can get all the schemas you need, for all, one or several properties - you need to pass the array of objects with properties `name` and `schema` of data properties that you want to validate.

**Comment by mbroadst** [comment]
@epoberezkin are you simply against supporting partial validation from a an absolute standpoint, or is it that the implementation would be very difficult? Would you be willing to accept support for this feature if I were to attempt an implementation? 

While you are correct that its _possible_ to support this functionality with the aforementioned solutions, this amount of working around the problem seems a bit much. Not only do I need to programmatically split up the existing schema into new one-off schemas, but those schemas themselves aren't even really "correct" in that I can't even validate the individual property anymore, I need to create a whole new object and superficially add a single property to it just to validate using the rules already present in the previously compiled validator.

It just seems like the original compiled validator should already have all the tools I need for what I'm trying to accomplish, but I'm sure you're coming from a far more informed position in your resistance to the feature.

**Comment by epoberezkin** [comment]
I understand that what you want is for a validator to somehow derive the requirement for a single property from a general schema that has the requirement for the whole object. Please correct me if I am wrong.

The problems here are:
- it is only applicable to objects, while schemas can be defined for other types
- it is not always possible - there are schemas where the requirements to a property are dependent on other properties
- it is complex even in relatively simple cases when it is possible (it should take into account patternProperties, required, schemas inside additionalProperties, compound keywords, etc.) 
- it is out of scope of what a validator should be doing. You want "given schema A derive schema B". Such schema operations are out of scope of validation.

What you want is only possible for schemas that are defined for objects, with properties independent of each other.

What I suggested above are not workarounds, these are generic approaches that would work for any schemas, without limiting them to a subset of JSON-schemas. These approaches are the opposite of what you want - instead of trying to derive the schema for a property from the schema for the object (complex, not always possible), I suggest to start from multiple schemas for all properties and to construct (manually or programmatically) the schema for the whole object (simple, always possible).

There is a package [ajv-inspector](https://github.com/e0ipso/ajv-inspector) by @e0ipso which seems to be doing something close to what you want: given the schema for an object it extracts the schema for a given property. There are several relatively simple improvements that would make it more useful - I've submitted several issues there, please also see https://github.com/epoberezkin/ajv/issues/166#issuecomment-239929160 .

I disagree that any such schema transformations/operations should be a part of Ajv. They should be separate packages. So please contribute to ajv-inspector if it is close to what you need.

**Comment by sengkathir** [comment]
@epoberezkin, sorry to comment in the same thread. Could you explain, in your template example

```
{
  "id": "someSchema",
  "type": "object",
  "properties": {
    {{~ it.properties:p:i }}
      {{?i}},{{?}}
      "{{= p.name }}": {{= JSON.stringify(p.schema) }}
    {{~}}
  },
  "required": [
    {{~ it.properties:p:i }}
      {{?i}},{{?}}
      "{{= p.name }}"
    {{~}}
  ]
}
```

`p` is 

```
{
  "name": "email",
  "schema": { "type": "string", "format": "email" }
}
```

or 

```
{
  "name": "email",
  "schema": {
      "id": "#email",
      "properties": {
        "email": { "type": "string", "format": "email" }
      },
      "required": [ "email" ]
    }
}
```
?

**Comment by sengkathir** [comment]
@epoberezkin, following data structure is working

```
{
  "name": "email",
  "schema": { "type": "string", "format": "email" }
}
```

is there any way to use fields with following data structure in dynamically generated templates?
```
{
  "name": "email",
  "schema": {
      "id": "#email",
      "properties": {
        "email": { "type": "string", "format": "email" }
      },
      "required": [ "email" ]
    }
}
```

thanks!

**Comment by epoberezkin** [comment]
The second can be created from the first.

**Comment by iamdanthedev** [comment]
@mbroadst Hi, I've just hit the same problem (validating just one field). 
In my project I am using only a subset of json schema (only objects, no dependencies, etc)

It seems like in order to validate a single field you can simply tear out it's property out of the whole schema and provide to a validator

In the example below this._schema points to the json schema object and this._db.$ajv is a mere instance of Ajv

```typescript
  public validateProp(prop: string, value: any): boolean {
    // const valid = this._validate(value, `.${prop}`);
    const valid = this._db.$ajv.validate(this._schema.properties[prop], value);

    if (valid === true) {
      return true;
    }
    else if (this._db.$ajv.errors) {
      throw new ValidationError(this._db.$ajv.errors);
    }
    else {
      throw new Error('Unknown error occured');
    }
  }
```

**Comment by iamdanthedev** [comment]
The only thing, I don't know whether ajv will compile a new schema on every run of this function for a single property. Maybe I should provide an $id or $ref to avoid recompilation? Maybe @epoberezkin could gracefully comment on this matter?

Thanks

**Comment by epoberezkin** [comment]
@rasdaniil Ajv doesn't compile the schema every time in this case, it stable-stringifies the JSON and looks up the compiled schema in the cache.

It is more efficient though to use `getSchema` (or to pass the jsonpointer to `ajv.validate`), as there is no need to stringify the schema in this case:

```javascript
// this should be executed only once
ajv.addSchema(schema); // schema should have $id or id attribute, say "schema.json#"
// or
// ajv.addSchema(schema, 'schema.json');

// ...
// every time you validate
const schemaUri = 'schema.json#/properties/' + prop;
const validate = ajv.getSchema(schemaUri); // you can keep them in the map, to avoid getSchema call too
const valid = validate(value); // errors are in validate.errors
// or
// const valid = ajv.validate(schemaUri, value); // errors are in ajv.errors
```

**Comment by iamdanthedev** [comment]
Thanks for quick response!

**Comment by FossPrime** [comment]
@epoberezkin it seems most people want to check just one property.

I'm trying to validate a patch call, which may have 20 different properties out of 50. I would really love an AJV option that ignores `required`

The way im working around it is duplicating the schema in memory and adding `required` to the schema on the copy. 

workaround feathers psudo code
```
partialSchema =  { ... }
schema = Object.assign({}, partialSchema, {required: ['name', 'email']})
hooks = {
  create: validateSchema(schema, ajv)
  update: validateSchema(schema, ajv)
  patch: validateSchema(partialSchema, ajv) 
}
```

**Comment by vladyslav2** [comment]
@epoberezkin still no update for this issue?

ajv looks like a nice package but it does not fit in restful patch/put methods

**Comment by agg23** [comment]
I'm another interested in using AJV for validating PATCHmethods, but any schema that has required properties breaks for that use case. My current hack is to duplicate the current schema, remove the required property from each schema object, then add this new schema to AJV:

```
ajvValidator.addSchema(schema, "schema");
ajvValidator.addSchema(partialSchema(JSON.parse(JSON.stringify(schema))), "partialSchema");

function partialSchema(dbschema: any) {
    let partialSchema = Object.assign({}, dbschema);
    for(let schemaName in partialSchema.definitions) {
        let schemaEntry = partialSchema.definitions[schemaName];
        if("required" in schemaEntry) {
            delete schemaEntry.required;
        }
    }

    return partialSchema;
}
```

This can then be referenced by:
`ajvValidator.validate("schema#/definitions/Item", request.body)` for POST or
`ajvValidator.validate("partialSchema#/definitions/Item", request.body)` for PATCH.

**Comment by vladyslav2** [comment]
@agg23 
> My current hack is to duplicate the current schema, remove the required property from each schema object, then add this new schema to AJV

Do not forget about a case when one field depends on other fields so you need to check if a user passed field1 then field2 should become required

**Comment by ghost** [comment]
```

const partialRequest = (schema: any, definition: string) => ({
  ...schema,
  definitions: {
    ...schema.definitions,
    [definition]: {
      ...schema.definitions[definition],
      required: [],
    },
  },
});

export const validateObject = (
  obj: any,
  definition: string,
  schema: string | object,
  withActivityStatus = true,
) => {
  const schemaObj =
    typeof schema === "string" ? getObjectFromFile<any>(schema) : schema;

  const validateSchema = withActivityStatus
    ? partialRequest(schemaObj, definition)
    : schemaObj;

  const validate = ajv.compile({
    ...validateSchema,
    $ref: `#/definitions/${definition}`,
  });
  const valid = validate(obj);
  if (!valid) {
    log.warn(
      `Validation failed for ${definition}, ${JSON.stringify(validate.errors)}`,
    );
  }
  const messages = validate.errors?.map((error) => `${error.message} ${error}`);
  return {
    valid,
    schema: schemaObj,
    errorMessage: JSON.stringify(messages),
    errors: validate.errors,
  };
};
```

**Comment by Pithikos** [comment]
Instead of creating complex and error-prone subschemas, I think a simpler solution for forms, is to simply always do full validation and then just choose to populate errors related to your specific field.

Example (not tested):
```javascript
const errors = []

function validateField(fieldName){
    const valid = ajvValidate(allValues);
    if (!valid) {
        for (const error of ajvValidate.errors) {
            if (error.instancePath.length === 0) {
                continue;
            }
            const propertyName = error.instancePath.slice(1);
            if (propertyName === fieldName) {
                errors.push(error.message);
            }
        }
    }
}
```

I assume that since AJV compiles the schema at the start, there's minimal performance hit. But feel free to correct me

**Comment by DesignByOnyx** [comment]
@Pithikos gave a real clever way of doing this. I wrote function essentially ignores "required" validators - effectively mimicking the TS `Partial<T>` utility. I haven't done thorough testing with complex data structures, custom formats & keywords, etc. But my initial testing works like a charm.

```ts
/**
 * Given a validator function and data, only validates the fields present on the data.
 * This essentially removes the "required" validation while still enforcing all other validation.
 * This has the same exact usage as a normal AsyncValidateFunction.
 */
const validatePartial = async <T>(validator: AsyncValidateFunction<T>, data: unknown) => {
  try {
    await validator(data);
  } catch (err) {
    if (!(err instanceof ValidationError)) throw err;

    const partialErrors = err.errors.filter((e) => e.keyword !== 'required');
    if (partialErrors.length) throw new ValidationError(partialErrors);
  }

  return data;
};
```

Now you should be able to validate partial data structures:

```ts
const validate = ajv.compile({
  $async: true,
  type: 'object',
  properties: {
    foo: { type: 'string' },
    bar: { type: 'number' },
    bing: { type: 'boolean' },
  },
  required: ['foo', 'bar'],
});

await validatePartial(validate, { bar: 1234 }); //-> valid!
await validatePartial(validate, { bing: true }); //-> valid!
await validatePartial(validate, { bing: 'asdf' }); //-> FAIL!
```
